### ISSUE TITLE
Usage Question

### ISSUE THREAD

**Issue by java8964** [nielsbasjes/logparser#40]
I am testing with version 3.0 and I have the Apache log format like following:
log.format=%a %{Host}i %u %t \"%r\" %>s %O \"%{Referer}i\" \"%{User-Agent}i\" %{Content-length}i %P %A

It will generate following lists based on example given here: https://github.com/nielsbasjes/logparser/blob/master/README-Java.md

STRING:connection.client.user
IP:connection.server.ip
TIME.STAMP:request.receive.time.last
TIME.DAY:request.receive.time.last.day
TIME.MONTHNAME:request.receive.time.last.monthname
TIME.MONTH:request.receive.time.last.month
TIME.WEEK:request.receive.time.last.weekofweekyear
TIME.YEAR:request.receive.time.last.weekyear
TIME.YEAR:request.receive.time.last.year
TIME.HOUR:request.receive.time.last.hour
TIME.MINUTE:request.receive.time.last.minute
TIME.SECOND:request.receive.time.last.second
TIME.MILLISECOND:request.receive.time.last.millisecond
TIME.DATE:request.receive.time.last.date
TIME.TIME:request.receive.time.last.time
TIME.ZONE:request.receive.time.last.timezone
TIME.EPOCH:request.receive.time.last.epoch
TIME.DAY:request.receive.time.last.day_utc
TIME.MONTHNAME:request.receive.time.last.monthname_utc
TIME.MONTH:request.receive.time.last.month_utc
TIME.WEEK:request.receive.time.last.weekofweekyear_utc
TIME.YEAR:request.receive.time.last.weekyear_utc
TIME.YEAR:request.receive.time.last.year_utc
TIME.HOUR:request.receive.time.last.hour_utc
TIME.MINUTE:request.receive.time.last.minute_utc
TIME.SECOND:request.receive.time.last.second_utc
TIME.MILLISECOND:request.receive.time.last.millisecond_utc
TIME.DATE:request.receive.time.last.date_utc
TIME.TIME:request.receive.time.last.time_utc
HTTP.URI:request.referer
HTTP.PROTOCOL:request.referer.protocol
HTTP.USERINFO:request.referer.userinfo
HTTP.HOST:request.referer.host
HTTP.PORT:request.referer.port
HTTP.PATH:request.referer.path
HTTP.QUERYSTRING:request.referer.query
STRING:request.referer.query.*
HTTP.REF:request.referer.ref
TIME.STAMP:request.receive.time
TIME.DAY:request.receive.time.day
TIME.MONTHNAME:request.receive.time.monthname
TIME.MONTH:request.receive.time.month
TIME.WEEK:request.receive.time.weekofweekyear
TIME.YEAR:request.receive.time.weekyear
TIME.YEAR:request.receive.time.year
TIME.HOUR:request.receive.time.hour
TIME.MINUTE:request.receive.time.minute
TIME.SECOND:request.receive.time.second
TIME.MILLISECOND:request.receive.time.millisecond
TIME.DATE:request.receive.time.date
TIME.TIME:request.receive.time.time
TIME.ZONE:request.receive.time.timezone
TIME.EPOCH:request.receive.time.epoch
TIME.DAY:request.receive.time.day_utc
TIME.MONTHNAME:request.receive.time.monthname_utc
TIME.MONTH:request.receive.time.month_utc
TIME.WEEK:request.receive.time.weekofweekyear_utc
TIME.YEAR:request.receive.time.weekyear_utc
TIME.YEAR:request.receive.time.year_utc
TIME.HOUR:request.receive.time.hour_utc
TIME.MINUTE:request.receive.time.minute_utc
TIME.SECOND:request.receive.time.second_utc
TIME.MILLISECOND:request.receive.time.millisecond_utc
TIME.DATE:request.receive.time.date_utc
TIME.TIME:request.receive.time.time_utc
HTTP.URI:request.referer.last
HTTP.PROTOCOL:request.referer.last.protocol
HTTP.USERINFO:request.referer.last.userinfo
HTTP.HOST:request.referer.last.host
HTTP.PORT:request.referer.last.port
HTTP.PATH:request.referer.last.path
HTTP.QUERYSTRING:request.referer.last.query
STRING:request.referer.last.query.*
HTTP.REF:request.referer.last.ref
NUMBER:connection.server.child.processid
BYTES:response.bytes
BYTESCLF:response.bytes
HTTP.HEADER:request.header.content-length
IP:connection.client.ip.last
HTTP.USERAGENT:request.user-agent.last
STRING:request.status.last
HTTP.USERAGENT:request.user-agent
STRING:connection.client.user.last
IP:connection.client.ip
HTTP.HEADER:request.header.host
HTTP.FIRSTLINE:request.firstline.original
HTTP.METHOD:request.firstline.original.method
HTTP.URI:request.firstline.original.uri
HTTP.PROTOCOL:request.firstline.original.uri.protocol
HTTP.USERINFO:request.firstline.original.uri.userinfo
HTTP.HOST:request.firstline.original.uri.host
HTTP.PORT:request.firstline.original.uri.port
HTTP.PATH:request.firstline.original.uri.path
HTTP.QUERYSTRING:request.firstline.original.uri.query
STRING:request.firstline.original.uri.query.*
HTTP.REF:request.firstline.original.uri.ref
HTTP.PROTOCOL_VERSION:request.firstline.original.protocol
HTTP.PROTOCOL:request.firstline.original.protocol
HTTP.PROTOCOL.VERSION:request.firstline.original.protocol.version
BYTES:response.bytes.last
BYTESCLF:response.bytes.last
HTTP.FIRSTLINE:request.firstline
HTTP.METHOD:request.firstline.method
HTTP.URI:request.firstline.uri
HTTP.PROTOCOL:request.firstline.uri.protocol
HTTP.USERINFO:request.firstline.uri.userinfo
HTTP.HOST:request.firstline.uri.host
HTTP.PORT:request.firstline.uri.port
HTTP.PATH:request.firstline.uri.path
HTTP.QUERYSTRING:request.firstline.uri.query
STRING:request.firstline.uri.query.*
HTTP.REF:request.firstline.uri.ref
HTTP.PROTOCOL_VERSION:request.firstline.protocol
HTTP.PROTOCOL:request.firstline.protocol
HTTP.PROTOCOL.VERSION:request.firstline.protocol.version
NUMBER:connection.server.child.processid.last
IP:connection.server.ip.last

Here is one sample log:
184.105.247.196 - - [03/Apr/2017:03:27:28 -0600] "GET /havimusor/month.calendar/2006/10/28/67.html HTTP/1.1" 404 419 "-" "-" - 115052 173.254.111.153

And here is my test code:
```
public class LogRecord {
    String clientIp;
    String customerHostName;
    String remoteUser;
    String dateTimeS;
    String httpRequest;
    String httpStatusCode;
    String bytesSent;
    String url;
    String userAgent;
    String contentLength;
    String serverProcessId;
    String customerDedicatedIP;

    public String getClientIp() {
        return clientIp;
    }

    @Field("IP:connection.client.ip")
    public void setClientIp(String clientIp) {
        this.clientIp = clientIp;
    }

    public String getCustomerHostName() {
        return customerHostName;
    }

    @Field("HTTP.HEADER:request.header.host")
    public void setCustomerHostName(String customerHostName) {
        this.customerHostName = customerHostName;
    }

    public String getRemoteUser() {
        return remoteUser;
    }

    @Field("STRING:connection.client.user")
    public void setRemoteUser(String remoteUser) {
        this.remoteUser = remoteUser;
    }

    public String getDateTimeS() {
        return dateTimeS;
    }

    @Field("TIME.STAMP:request.receive.time")
    public void setDateTimeS(String dateTimeS) {
        this.dateTimeS = dateTimeS;
    }

    public String getHttpRequest() {
        return httpRequest;
    }

    @Field("HTTP.FIRSTLINE:request.firstline")
    public void setHttpRequest(String httpRequest) {
        this.httpRequest = httpRequest;
    }

    public String getHttpStatusCode() {
        return httpStatusCode;
    }

    @Field("STRING:request.status.last")
    public void setHttpStatusCode(String httpStatusCode) {
        this.httpStatusCode = httpStatusCode;
    }

    public String getBytesSent() {
        return bytesSent;
    }

    @Field("BYTES:response.bytes")
    public void setBytesSent(String bytesSent) {
        this.bytesSent = bytesSent;
    }

    public String getUrl() {
        return url;
    }

    @Field("HTTP.URI:request.referer")
    public void setUrl(String url) {
        this.url = url;
    }

    public String getUserAgent() {
        return userAgent;
    }

    @Field("HTTP.USERAGENT:request.user-agent")
    public void setUserAgent(String userAgent) {
        this.userAgent = userAgent;
    }

    public String getContentLength() {
        return contentLength;
    }

    @Field("HTTP.HEADER:request.header.content-length")
    public void setContentLength(String contentLength) {
        this.contentLength = contentLength;
    }

    public String getServerProcessId() {
        return serverProcessId;
    }

    @Field("NUMBER:connection.server.child.processid")
    public void setServerProcessId(String serverProcessId) {
        this.serverProcessId = serverProcessId;
    }

    public String getCustomerDedicatedIP() {
        return customerDedicatedIP;
    }

    @Field("IP:connection.server.ip.last")
    public void setCustomerDedicatedIP(String customerDedicatedIP) {
        this.customerDedicatedIP = customerDedicatedIP;
    }

    public LogRecord() { }
}
```

Here is the issue for the httpRequest field. I may have some log entries like following example:
184.105.247.196 - - [03/Apr/2017:03:27:28 -0600] **"\x16\x03\x01"** 404 419 "-" "-" - 115052 173.254.111.153

And the above code will throw following error:
The input line does not match the specified log format.Line     : Value{filled=STRING, s='184.105.247.196 - - [03/Apr/2017:03:27:28 -0600] "\x16\x03\x01" 404 419 "-" "-" - 115052 173.254.111.153', l=null, d=null}
LogFormat: %a %{Host}i %u %t "%r" %>s %O "%{Referer}i" "%{User-Agent}i" %{Content-length}i %P %A
RegEx    : ^((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|:?(?:[0-9a-fA-F]{1,4}(?::|.)?){0,8}(?::|::)?(?:[0-9a-fA-F]{1,4}(?::|.)?){0,8}|-)\Q \E(.*)\Q \E(.*)\Q [\E([0-3][0-9]/(?:[a-zA-Z][a-zA-Z][a-zA-Z])/[1-9][0-9][0-9][0-9]:[0-9][0-9]:[0-9][0-9]:[0-9][0-9] [\+|\-][0-9][0-9][0-9][0-9])\Q] "\E((?:[a-zA-Z-_]+ .*(?: HTTP/[0-9]+\.[0-9]+)?)|-)\Q" \E([^\s]*)\Q \E([0-9]*|-)\Q "\E(.*)\Q" "\E(.*)\Q" \E(.*)\Q \E([0-9]*)\Q \E((?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|:?(?:[0-9a-fA-F]{1,4}(?::|.)?){0,8}(?::|::)?(?:[0-9a-fA-F]{1,4}(?::|.)?){0,8}|-)$

The problem is that I almost tried all other format, with following:
STRING:request.firstline.uri.query.*
STRING:request.firstline.original.uri.query.*
HTTP.FIRSTLINE:request.firstline.original
etc

Some of them won't produce error, but they cannot catch the good request any more, like "GET /havimusor/month.calendar/2006/10/28/67.html HTTP/1.1" in the example.

I wonder what format I should specified to get both "GET /havimusor/month.calendar/2006/10/28/67.html HTTP/1.1" and "\x16\x03\x01" in this field.

2nd question is that we have 12 fields in the log, for now, I just want to catch these 12 fields. We have another implementation using REGEX, but as the data cases complex enough, I like to reuse your library. The issue is that I tested with 2.1M rows, but the performance is about 10x difference. Looks like internally you also use REGEX. I wonder if any place I misuse your library, and lead to such performance difference.

Here is my sample test code of using the above log record object:

```
        BufferedReader br = new BufferedReader(new FileReader(args[0]));
        String line;
        long totalRecordCnt = 0;
        long starting = System.currentTimeMillis();
        while ((line = br.readLine()) != null) {
            totalRecordCnt++;
            try {
                record = parser.parse(record, line);
            } catch (Throwable t) {
                System.err.println(t.getMessage());
            }
        }
        br.close();
        System.out.println("Parsing took " + (System.currentTimeMillis() - starting) + " mills, with " + totalRecordCnt + " total records!");
```
For this library, it took around 160s to 200s for 2146891 (68MB) records, and using plan REGEX to parse the same data will take about 13s to 15s range. I want to confirm what I did is the correct way to use this library, and **I DO want to get all 12 fields out**, as this logic is in our Spark ETL.

**Comment by nielsbasjes** [comment]
1) The root cause of the parse error is that in the regex I generate I expect the firstline to look like this: 
`(?:[a-zA-Z-_]+ .*(?: HTTP/[0-9]+\\.[0-9]+)?)|-`

The value \x16\x03\x01 does not match that pattern so the parsing for that line fails.
I just had a look and I think I fixed it (it now allows the firstline to be `.*`).
Can you verify please?

2) The code you posted looks fine and is how I intended it to be used.
I ran this code by simply parsing the same line 2146891 times and this took approx 48 seconds on my laptop (i7). So yes, slower but not that slow.

Yes, internally I generate a regex for the first parsing layer. 
You got to see it as part of the error in the previous question.

The regex is created as 'tightly as possible'. So for example for the IP address the regex is quite elaborate because I want to trip the 'wrong format' situation an still allow both ipv4 and ipv6. This system is slower than 'just a regex' and perhaps there is room to make it faster. Yet this is one of those balances between 'ease of use' vs 'maximum speed'.

**Comment by nielsbasjes** [comment]
One remark on the way you use the code: The parser will ONLY call the setter if it actually has the field.
So if you extract a field that is sometimes present and sometimes it is not you may have 'old values' in your record.
I recommend you include a method to clear the values before you pass it again into the parser.

If for example I add this
```       String method;
        @Field("HTTP.METHOD:request.firstline.method")
        public void setMethod(String method) {
            this.method = method;
        }
```
And then first call it with the 'normal firstline' and the with the  "\x16\x03\x01" I would see in both records the method being "GET". In the first record it was extracted, in the second record it was not there so nothing was set and the old value wat not overwritten.

**Comment by java8964** [comment]
Yes. This issue is gone in version of 3.1-SNAPSHOT from master. When do you plan to release it? For the performance, we may use this library as a additional logic to parse the data if the plan REGEX cannot handle it, so it won't impact the performance too much

**Comment by nielsbasjes** [comment]
I just released version 3.1 that includes these changes.
Should be in maven central in a few hours or so.
